File: java/src/Alfredo.java
package pastas;

public class Alfredo extends Pasta {
	private String _addOn;
	private boolean _cached = false;
	private String _dish = null;
	
	public Alfredo(String addOn) {
		_addOn = addOn;
	}
	public Alfredo(){
		this("Plain");
	}
	
	private void _updateCache() {
		if(_addOn == "Plain"){
			_dish = "Classic Alfredo";
			_cached = true;
		}
		else {
			_dish = _addOn + " Alfredo";
			_cached = true;
		}
	}
	
	public String getAddOn(){
		return _addOn;
	}
	
	public void setAddOn(String value) {
		if(value != _addOn) {
			_addOn = value;
			_cached = false;
		}
	}
	
	public String getDish() {
		if(!_cached) {
			_updateCache();
		}
		return _dish;
	}
}

File: CPP/src/Alfredo.cpp
#include "Alfredo.h"

namespace pastas {
	Alfredo::Alfredo(const std::string &addOn)
		:_addOn(addOn),_dish("Alfredo"), _cheese("Asiago")
	{
	}
	
	std::string Alfredo::getAddOn() const {
		return _addOn;
	}
	
	void Alfredo::setAddOn(const std::string &value) {
		if(_addOn != value) {
			_addOn = value;
			updateDish();
		}
	}
	
	void Alfredo::updateDish(){
		_dish = _addOn.append(" Alfredo");
	}
	
	std::string Alfredo::getDish() const {
		return _dish;
	}
}